R/hcai-impute.R

```diff
@@ -0,0 +1,133 @@
+#' @param nominal_method Defaults to \code{"new_category"}. Other choices are 
+#' \code{"bagimpute"} or \code{"knnimpute"}. 
+#' @param numeric_params A named list with parmeters to use with chosen
+#' imputation method on numeric data. Options are \code{bag_model}, 
+#' \code{bag_options}, \code{knn_k}, \code{impute_with}, or \code{seed_val}. 
+#' See \link{step_bagimpute} or \link{step_knnimpute} for details.
+#' @param nominal_params A named list with parmeters to use with chosen
+#' imputation method on nominal data. Options are \code{bag_model}, 
+#' \code{bag_options}, \code{knn_k}, \code{impute_with}, or \code{seed_val}. 
+#' See \link{step_bagimpute} or \link{step_knnimpute} for details.
+#' @return An updated version of `recipe` with the new step
+#'  added to the sequence of existing steps.
+#'
+#' @export
+#' @import recipes
+
+#' @examples
+#' library(healthcareai)
+#' library(recipes)
+#' 
+#' n = 100
+#' d <- tibble::tibble(encounter_id = 1:n,
+#'             patient_id = sample(1:20, size = n, replace = TRUE),
+#'             hemoglobin_count = rnorm(n, mean = 15, sd = 1),
+#'             hemoglobin_category = sample(c("Low", "Normal", "High", NA), 
+#'                                          size = n, replace = TRUE),
+#'             disease = ifelse(hemoglobin_count < 15, "Yes", "No")
+#' )
+#' 
+#' # Initialize
+#' my_recipe <- recipe(disease ~ ., data = d)
+#' 
+#' # Create recipe
+#' my_recipe <- my_recipe %>%
+#'   hcai_impute(numeric_method = "mean",
+#'     nominal_method = "new_category")
+#' my_recipe
+#' 
+#' # Train recipe
+#' trained_recipe <- prep(my_recipe, training = d)
+#' 
+#' # Apply recipe
+#' data_modified <- bake(trained_recipe, newdata = d)
+#' countMissingData(data_modified)
+#' 
+hcai_impute <- function(rec_obj,
+                        numeric_method = "mean",
+                        nominal_method = "new_category",
+                        numeric_params = NULL,
+                        nominal_params = NULL) {
+  # Check to make sure rec_obj is the right type
+  if (class(rec_obj) != "recipe") {
+    stop("rec_obj must be recipe object"
+    )
+  }
+
+  # If methods dont exist, use defaults
+  if (!(numeric_method %in% c("mean", "bagimpute", "knnimpute"))) {
+    stop("non-supported numeric method. Use \"mean\", \"bagimpute\",
+         or \"knnimpute\"")
+  }
+  if (!(nominal_method %in% c("new_category", "bagimpute", "knnimpute"))) {
+    stop("non-supported nominal method. Use \"new_category\", \"bagimpute\",
+         or \"knnimpute\"")
+  }
+
+  # Assign defaults for params
+  defaults <- list(
+    bag_model = NULL,
+    bag_options = list(nbagg = 25, keepX = FALSE),
+    knn_K = 5,
+    impute_with = imp_vars(all_predictors()),
+    seed_val = sample.int(1000, 1)
+  )
+
+  # Fill in user-specified params
+  num_p <- defaults
+  num_p[names(num_p) %in% names(numeric_params)] <- numeric_params
+  nom_p <- defaults
+  nom_p[names(nom_p) %in% names(nominal_params)] <- nominal_params
```

Comment:
This is elegant! I think we ought to at least warn if the user specifies a parameter that doesn't end up getting used. E.g. If the user provides a knn parameter to numeric_params but doesn't specify numeric_method = "knn"

Reply:
Just for funzies, you can do `num_p <- nom_p <- defaults`. Maybe it's a little clearer too.

Reply:
I added this warning. I choose to see if there are any extra params, regardless of the method. This means that a user could use `bagimpute` and `knn_K` without getting a warning, but it saves quite a lot of code to just check for things like, `bogus_parameter` and then provide a list of the accepted ones.

Multiple arrows? Tricksy...